modules/geoip2/maxminddb-helper.c

```diff
@@ -0,0 +1,355 @@
+      break;
+    case MMDB_DATA_TYPE_UINT16:
+      g_string_append_printf(target, "%u", entry_data->uint16);
+      break;
+    case MMDB_DATA_TYPE_UINT32:
+      g_string_append_printf(target, "%u", entry_data->uint32);
+      break;
+    case MMDB_DATA_TYPE_INT32:
+      g_string_append_printf(target, "%d", entry_data->int32);
+      break;
+    case MMDB_DATA_TYPE_UINT64:
+      g_string_append_printf(target, "%lu", entry_data->uint64);
+      break;
+    case MMDB_DATA_TYPE_BOOLEAN:
+      g_string_append_printf(target, "%s", entry_data->boolean ? "true" : "false");
+      break;
+    case MMDB_DATA_TYPE_UINT128:
+    case MMDB_DATA_TYPE_MAP:
+    case MMDB_DATA_TYPE_BYTES:
+    case MMDB_DATA_TYPE_ARRAY:
+      g_assert_not_reached();
+    default:
+      g_assert_not_reached();
+    }
+}
+
+static MMDB_entry_data_list_s *
+mmdb_skip_tree(MMDB_entry_data_list_s *entry_data_list, gint *status)
+{
+
+  switch (entry_data_list->entry_data.type)
+    {
+    case MMDB_DATA_TYPE_MAP:
+    {
+      guint32 size = entry_data_list->entry_data.data_size;
+
+      for (entry_data_list = entry_data_list->next;
+           size && entry_data_list; size--)
+        {
+
+          if (MMDB_DATA_TYPE_UTF8_STRING != entry_data_list->entry_data.type)
+            {
+              *status = MMDB_INVALID_DATA_ERROR;
+              return NULL;
+            }
+          entry_data_list = entry_data_list->next;
+          entry_data_list = mmdb_skip_tree(entry_data_list, status);
+          if (MMDB_SUCCESS != *status)
+            {
+              return NULL;
+            }
+        }
+    }
+    break;
+    case MMDB_DATA_TYPE_ARRAY:
+    {
+      uint32_t size = entry_data_list->entry_data.data_size;
+
+      for (entry_data_list = entry_data_list->next;
+           size && entry_data_list; size--)
+        {
+          entry_data_list = mmdb_skip_tree(entry_data_list, status);
+
+          if (MMDB_SUCCESS != *status)
+            {
+              return NULL;
+            }
+        }
+    }
+    break;
+    case MMDB_DATA_TYPE_BYTES:
+    case MMDB_DATA_TYPE_BOOLEAN:
+    case MMDB_DATA_TYPE_UINT128:
+    case MMDB_DATA_TYPE_UTF8_STRING:
+    case MMDB_DATA_TYPE_DOUBLE:
+    case MMDB_DATA_TYPE_FLOAT:
+    case MMDB_DATA_TYPE_UINT16:
+    case MMDB_DATA_TYPE_UINT32:
+    case MMDB_DATA_TYPE_UINT64:
+    case MMDB_DATA_TYPE_INT32:
+      entry_data_list = entry_data_list->next;
+      break;
+
+    default:
+      *status = MMDB_INVALID_DATA_ERROR;
+      return NULL;
+    }
+
+  *status = MMDB_SUCCESS;
+  return entry_data_list;
+}
+
+static void
+_geoip_log_msg_add_value(LogMessage *msg, GArray *path, GString *value)
+{
+  gchar *path_string = g_strjoinv(".", (gchar **)path->data);
+  log_msg_set_value_by_name(msg, path_string, value->str, value->len);
+  g_free(path_string);
+}
+
+static MMDB_entry_data_list_s *
+select_language(LogMessage *msg, MMDB_entry_data_list_s *entry_data_list, GArray *path, gint *status)
```

Comment:
I could not find a way to repattern it as general as a filter with predicate and callback, because in this special case both predicate (key == en) and callback (i.e insert data to nvtable) needs the same variable: the key, and I would like to avoid to calculate the key twice.

But I could extract 3 functions here:
return_and_set_error_if
check_language_and_maybe_insert
which latter encapsulates in the end:
_print_preferred_string_for_lang